### PR TITLE
Fix the often call of event tox_friend_connection_status

### LINKED ISSUE #868: A tox_friend_connection_status event often occurs
I began to notice that when my client (ToxPhone) is running, a tox_friend_connection_status event occurs quite often. 

```
12.04.2018 10:31:55.944846 DEBUG   LWP22452 [tox_net.cpp:1022 ToxNet] ToxEvent: friend TCP connection. Friend name/number: hkarel/7
12.04.2018 10:31:57.790578 DEBUG   LWP22452 [tox_net.cpp:1022 ToxNet] ToxEvent: friend UPD connection. Friend name/number: hkarel/7
12.04.2018 10:32:04.934513 DEBUG   LWP22452 [tox_net.cpp:1022 ToxNet] ToxEvent: friend TCP connection. Friend name/number: hkarel/7
12.04.2018 10:32:06.798105 DEBUG   LWP22452 [tox_net.cpp:1022 ToxNet] ToxEvent: friend UPD connection. Friend name/number: hkarel/7
12.04.2018 10:32:13.915010 DEBUG   LWP22452 [tox_net.cpp:1022 ToxNet] ToxEvent: friend TCP connection. Friend name/number: hkarel/7
...
12.04.2018 12:09:54.783067 DEBUG   LWP22452 [tox_net.cpp:1022 ToxNet] ToxEvent: friend UPD connection. Friend name/number: hkarel/7
12.04.2018 12:10:00.795466 DEBUG   LWP22452 [tox_net.cpp:1022 ToxNet] ToxEvent: friend disonnected. Friend name/number: hkarel/7
12.04.2018 12:10:01.007814 DEBUG   LWP22452 [tox_net.cpp:1022 ToxNet] ToxEvent: friend UPD connection. Friend name/number: hkarel/7
12.04.2018 12:10:01.010377 DEBUG   LWP22452 [tox_net.cpp:931 ToxNet] ToxEvent: friend name changed. Friend name/number: hkarel/7
12.04.2018 12:10:01.010420 DEBUG   LWP22452 [tox_net.cpp:952 ToxNet] ToxEvent: friend status message changed. Friend name/number: hkarel/7
12.04.2018 12:57:02.663650 VERBOSE LWP22452 [tox_net.cpp:996 ToxNet] Disconnected from DHT
12.04.2018 12:57:02.876809 VERBOSE LWP22452 [tox_net.cpp:250 ToxNet] Connecting to bootstrap node 144.217.86.39 : 33445; name: velusip
12.04.2018 12:57:02.878354 VERBOSE LWP22452 [tox_net.cpp:250 ToxNet] Connecting to bootstrap node 2a00:7a60:0:746b::3 : 33445; name: MAH69K
12.04.2018 12:57:04.486442 VERBOSE LWP22452 [tox_net.cpp:250 ToxNet] Connecting to bootstrap node tox.abilinski.com : 33445; name: flobe
12.04.2018 12:57:05.770388 VERBOSE LWP22452 [tox_net.cpp:250 ToxNet] Connecting to bootstrap node tox.deadteam.org : 33445; name: DeadTeam
12.04.2018 12:57:08.480262 VERBOSE LWP22452 [tox_net.cpp:250 ToxNet] Connecting to bootstrap node node.tox.ngc.network : 33445; name: Nolz
12.04.2018 12:57:09.171811 VERBOSE LWP22452 [tox_net.cpp:250 ToxNet] Connecting to bootstrap node 2001:6f8:1c3c:babe::14:1 : 33445; name: Manolis
12.04.2018 12:57:09.670583 VERBOSE LWP22452 [tox_net.cpp:994 ToxNet] Connected to DHT through UDP
```
Generation of this event lasted an hour and a half with an interval of 10 seconds. And it stopped only after reconnecting to DHT. Full log is attached.
[toxphone.log.toxphone.txt](https://github.com/TokTok/c-toxcore/files/1902749/toxphone.log.toxphone.txt)

The output of the message "ToxEvent: friend TCP connection" is executed by the following code:
```
void ToxNet::tox_friend_connection_status(Tox* tox, uint32_t friend_number,
                                          TOX_CONNECTION connection_status, void* user_data)
{
    const char* stat;
    switch (connection_status)
    {
        case TOX_CONNECTION_TCP:
            stat = "TCP connection. ";
            break;

        case TOX_CONNECTION_UDP:
            stat = "UPD connection. ";
            break;

        default:
            stat = "disonnected. ";
    }
    log_debug_m << "ToxEvent: friend " << stat
                << ToxFriendLog(tox, friend_number);
    ...
}
```
I would like to understand why the event is generated so often and for a long time.

### PR DESCRIPTION
**Pull request by hkarel**
Fixes #868

### CODE REVIEW

**Comment by sudden6** [comment]
how did you determine that value?

**Comment by sudden6** [comment]
Looks solid :+1: 

I think you should also fix the condition to `>=` additionally, even if it works it's IMO better to do it right™

**Comment by hkarel** [comment]
> I think you should also fix the condition to >= additionally, even if it works it's IMO better to do it right™

Please clarify:
1. The condition '>=' use in conjunction with parameter UDP_DIRECT_TIMEOUT_OVERLAP?
2. Or leave only one condition '>=', and DP_DIRECT_TIMEOUT OVERLAP remove completely?

**Comment by sudden6** [comment]
use both, while the condition probably fixes the problem most of the time, we should also fix it for very slow connections.

**Comment by hkarel** [comment]
I did some research and got new information. 

1. The UDP packets are sent to the friend-node with the UDP_DIRECT_TIMEOUT/2 interval only until the UDP connection is not "established". After the UDP connection is "established", the confirmation packets are dispatched one time per second (this will be seen in the log below).
2. I decided to strengthen the logging to look at the intensity of the exchange of UDP packets of my client and [EchoBot](https://github.com/alexbakker/EchoBot).  EchoBot public key: 76518406... From the log you can see that the response from EchoBot can be 9 seconds, which is more than UDP_DIRECT_TIMEOUT (see entries `udp_handle_packet | PK: 76518406`). This is exactly what causes the original problem. At the same time, my client sends UDP packets to EchoBot with an interval of about 1 second (see `send_packet_to | PK: 76518406`). 
One might assume that the EchoBot is responding so slowly because of the client's high load, but the maximum latency interval is never more than 9 seconds. Therefore, I conclude that loading can be a factor, but there is something else. Do the main developers have any hypotheses about this?
I have a fear that if we accept these changes, the behavior of EchoBot may not change. Just the maximum delay interval will increase from 9 seconds to 11-12.
It would be great if the developer of the EchoBot agrees to build toxcore with this PR and conduct the testing.

```
29.04.2018 20:01:27.259727 DEBUG   LWP22584 [net_crypto.c:705 ToxCore] ---- send_packet_to    | PK: 43f22f1a9069eb041f5b11b4b6420e64 CRYPTO_DATA
29.04.2018 20:01:27.309881 DEBUG   LWP22584 [net_crypto.c:2525 ToxCore] --- udp_handle_packet | PK: 76518406f6a9f2217e8dc487cc783c25 CRYPTO_DATA 2
29.04.2018 20:01:27.360400 DEBUG   LWP22584 [net_crypto.c:705 ToxCore] ---- send_packet_to    | PK: 76518406f6a9f2217e8dc487cc783c25 CRYPTO_DATA
29.04.2018 20:01:27.462235 DEBUG   LWP22584 [net_crypto.c:705 ToxCore] ---- send_packet_to    | PK: 76518406f6a9f2217e8dc487cc783c25 CRYPTO_DATA
29.04.2018 20:01:27.562846 DEBUG   LWP22584 [net_crypto.c:705 ToxCore] ---- send_packet_to    | PK: f3c923e892339348fbba55d7fde7f368 CRYPTO_DATA
29.04.2018 20:01:27.613120 DEBUG   LWP22584 [net_crypto.c:2525 ToxCore] --- udp_handle_packet | PK: 4b1c717470f18181d00070da1a72f0fe CRYPTO_DATA 2
29.04.2018 20:01:27.613142 DEBUG   LWP22584 [net_crypto.c:2525 ToxCore] --- udp_handle_packet | PK: f3c923e892339348fbba55d7fde7f368 CRYPTO_DATA 2
29.04.2018 20:01:27.613206 DEBUG   LWP22584 [net_crypto.c:705 ToxCore] ---- send_packet_to    | PK: 4b1c717470f18181d00070da1a72f0fe CRYPTO_DATA
29.04.2018 20:01:27.764124 DEBUG   LWP22584 [net_crypto.c:2525 ToxCore] --- udp_handle_packet | PK: 43f22f1a9069eb041f5b11b4b6420e64 CRYPTO_DATA 2
29.04.2018 20:01:27.814577 DEBUG   LWP22584 [net_crypto.c:705 ToxCore] ---- send_packet_to    | PK: 649c778c3f8c3dcf69a840da130b2775 CRYPTO_DATA
29.04.2018 20:01:28.266178 DEBUG   LWP22584 [net_crypto.c:705 ToxCore] ---- send_packet_to    | PK: 43f22f1a9069eb041f5b11b4b6420e64 CRYPTO_DATA
29.04.2018 20:01:28.366443 DEBUG   LWP22584 [net_crypto.c:705 ToxCore] ---- send_packet_to    | PK: 76518406f6a9f2217e8dc487cc783c25 CRYPTO_DATA
29.04.2018 20:01:28.568122 DEBUG   LWP22584 [net_crypto.c:705 ToxCore] ---- send_packet_to    | PK: f3c923e892339348fbba55d7fde7f368 CRYPTO_DATA
29.04.2018 20:01:28.618259 DEBUG   LWP22584 [net_crypto.c:2525 ToxCore] --- udp_handle_packet | PK: 4b1c717470f18181d00070da1a72f0fe CRYPTO_DATA 2
29.04.2018 20:01:28.618326 DEBUG   LWP22584 [net_crypto.c:705 ToxCore] ---- send_packet_to    | PK: 4b1c717470f18181d00070da1a72f0fe CRYPTO_DATA
29.04.2018 20:01:28.668627 DEBUG   LWP22584 [net_crypto.c:2525 ToxCore] --- udp_handle_packet | PK: f3c923e892339348fbba55d7fde7f368 CRYPTO_DATA 2
29.04.2018 20:01:28.819036 DEBUG   LWP22584 [net_crypto.c:2525 ToxCore] --- udp_handle_packet | PK: 43f22f1a9069eb041f5b11b4b6420e64 CRYPTO_DATA 2
29.04.2018 20:01:28.819090 DEBUG   LWP22584 [net_crypto.c:705 ToxCore] ---- send_packet_to    | PK: 649c778c3f8c3dcf69a840da130b2775 CRYPTO_DATA
29.04.2018 20:01:29.270471 DEBUG   LWP22584 [net_crypto.c:705 ToxCore] ---- send_packet_to    | PK: 43f22f1a9069eb041f5b11b4b6420e64 CRYPTO_DATA
29.04.2018 20:01:29.370769 DEBUG   LWP22584 [net_crypto.c:705 ToxCore] ---- send_packet_to    | PK: 76518406f6a9f2217e8dc487cc783c25 CRYPTO_DATA
29.04.2018 20:01:29.572672 DEBUG   LWP22584 [net_crypto.c:705 ToxCore] ---- send_packet_to    | PK: f3c923e892339348fbba55d7fde7f368 CRYPTO_DATA
29.04.2018 20:01:29.622801 DEBUG   LWP22584 [net_crypto.c:2525 ToxCore] --- udp_handle_packet | PK: 4b1c717470f18181d00070da1a72f0fe CRYPTO_DATA 2
29.04.2018 20:01:29.622867 DEBUG   LWP22584 [net_crypto.c:705 ToxCore] ---- send_packet_to    | PK: 4b1c717470f18181d00070da1a72f0fe CRYPTO_DATA
29.04.2018 20:01:29.723194 DEBUG   LWP22584 [net_crypto.c:2525 ToxCore] --- udp_handle_packet | PK: f3c923e892339348fbba55d7fde7f368 CRYPTO_DATA 2
29.04.2018 20:01:29.823470 DEBUG   LWP22584 [net_crypto.c:2525 ToxCore] --- udp_handle_packet | PK: 43f22f1a9069eb041f5b11b4b6420e64 CRYPTO_DATA 2
29.04.2018 20:01:29.823560 DEBUG   LWP22584 [net_crypto.c:705 ToxCore] ---- send_packet_to    | PK: 649c778c3f8c3dcf69a840da130b2775 CRYPTO_DATA
29.04.2018 20:01:30.275106 DEBUG   LWP22584 [net_crypto.c:705 ToxCore] ---- send_packet_to    | PK: 43f22f1a9069eb041f5b11b4b6420e64 CRYPTO_DATA
29.04.2018 20:01:30.375662 DEBUG   LWP22584 [net_crypto.c:705 ToxCore] ---- send_packet_to    | PK: 76518406f6a9f2217e8dc487cc783c25 CRYPTO_DATA
29.04.2018 20:01:30.577015 DEBUG   LWP22584 [net_crypto.c:705 ToxCore] ---- send_packet_to    | PK: f3c923e892339348fbba55d7fde7f368 CRYPTO_DATA
29.04.2018 20:01:30.627164 DEBUG   LWP22584 [net_crypto.c:2525 ToxCore] --- udp_handle_packet | PK: 4b1c717470f18181d00070da1a72f0fe CRYPTO_DATA 2
29.04.2018 20:01:30.627234 DEBUG   LWP22584 [net_crypto.c:705 ToxCore] ---- send_packet_to    | PK: 4b1c717470f18181d00070da1a72f0fe CRYPTO_DATA
29.04.2018 20:01:30.777587 DEBUG   LWP22584 [net_crypto.c:2525 ToxCore] --- udp_handle_packet | PK: f3c923e892339348fbba55d7fde7f368 CRYPTO_DATA 2
29.04.2018 20:01:30.827776 DEBUG   LWP22584 [net_crypto.c:705 ToxCore] ---- send_packet_to    | PK: 649c778c3f8c3dcf69a840da130b2775 CRYPTO_DATA
29.04.2018 20:01:30.877924 DEBUG   LWP22584 [net_crypto.c:2525 ToxCore] --- udp_handle_packet | PK: 43f22f1a9069eb041f5b11b4b6420e64 CRYPTO_DATA 2
29.04.2018 20:01:31.179211 DEBUG   LWP22584 [net_crypto.c:2525 ToxCore] --- udp_handle_packet | PK: 43f22f1a9069eb041f5b11b4b6420e64 CRYPTO_DATA 2
29.04.2018 20:01:31.279558 DEBUG   LWP22584 [net_crypto.c:705 ToxCore] ---- send_packet_to    | PK: 43f22f1a9069eb041f5b11b4b6420e64 CRYPTO_DATA
29.04.2018 20:01:31.380087 DEBUG   LWP22584 [net_crypto.c:705 ToxCore] ---- send_packet_to    | PK: 76518406f6a9f2217e8dc487cc783c25 CRYPTO_DATA
29.04.2018 20:01:31.431022 DEBUG   LWP22584 [net_crypto.c:705 ToxCore] ---- send_packet_to    | PK: 4b1c717470f18181d00070da1a72f0fe CRYPTO_DATA
29.04.2018 20:01:31.431053 DEBUG   LWP22584 [net_crypto.c:705 ToxCore] ---- send_packet_to    | PK: 43f22f1a9069eb041f5b11b4b6420e64 CRYPTO_DATA
29.04.2018 20:01:31.531738 DEBUG   LWP22584 [net_crypto.c:2525 ToxCore] --- udp_handle_packet | PK: 4b1c717470f18181d00070da1a72f0fe CRYPTO_DATA 2
29.04.2018 20:01:31.581911 DEBUG   LWP22584 [net_crypto.c:2525 ToxCore] --- udp_handle_packet | PK: 43f22f1a9069eb041f5b11b4b6420e64 CRYPTO_DATA 2
29.04.2018 20:01:31.581937 DEBUG   LWP22584 [net_crypto.c:2525 ToxCore] --- udp_handle_packet | PK: 4b1c717470f18181d00070da1a72f0fe CRYPTO_DATA 2
29.04.2018 20:01:31.582037 DEBUG   LWP22584 [net_crypto.c:705 ToxCore] ---- send_packet_to    | PK: f3c923e892339348fbba55d7fde7f368 CRYPTO_DATA
29.04.2018 20:01:31.632258 DEBUG   LWP22584 [net_crypto.c:705 ToxCore] ---- send_packet_to    | PK: 4b1c717470f18181d00070da1a72f0fe CRYPTO_DATA
29.04.2018 20:01:31.833121 DEBUG   LWP22584 [net_crypto.c:2525 ToxCore] --- udp_handle_packet | PK: f3c923e892339348fbba55d7fde7f368 CRYPTO_DATA 2
29.04.2018 20:01:32.284857 DEBUG   LWP22584 [net_crypto.c:705 ToxCore] ---- send_packet_to    | PK: 43f22f1a9069eb041f5b11b4b6420e64 CRYPTO_DATA
29.04.2018 20:01:32.385300 DEBUG   LWP22584 [net_crypto.c:705 ToxCore] ---- send_packet_to    | PK: 76518406f6a9f2217e8dc487cc783c25 CRYPTO_DATA
29.04.2018 20:01:32.586672 DEBUG   LWP22584 [net_crypto.c:705 ToxCore] ---- send_packet_to    | PK: f3c923e892339348fbba55d7fde7f368 CRYPTO_DATA
29.04.2018 20:01:32.636811 DEBUG   LWP22584 [net_crypto.c:2525 ToxCore] --- udp_handle_packet | PK: 43f22f1a9069eb041f5b11b4b6420e64 CRYPTO_DATA 2
29.04.2018 20:01:32.636837 DEBUG   LWP22584 [net_crypto.c:2525 ToxCore] --- udp_handle_packet | PK: 4b1c717470f18181d00070da1a72f0fe CRYPTO_DATA 2
29.04.2018 20:01:32.636906 DEBUG   LWP22584 [net_crypto.c:705 ToxCore] ---- send_packet_to    | PK: 4b1c717470f18181d00070da1a72f0fe CRYPTO_DATA
29.04.2018 20:01:32.887617 DEBUG   LWP22584 [net_crypto.c:2525 ToxCore] --- udp_handle_packet | PK: f3c923e892339348fbba55d7fde7f368 CRYPTO_DATA 2
29.04.2018 20:01:33.289381 DEBUG   LWP22584 [net_crypto.c:705 ToxCore] ---- send_packet_to    | PK: 43f22f1a9069eb041f5b11b4b6420e64 CRYPTO_DATA
29.04.2018 20:01:33.389960 DEBUG   LWP22584 [net_crypto.c:705 ToxCore] ---- send_packet_to    | PK: 76518406f6a9f2217e8dc487cc783c25 CRYPTO_DATA
29.04.2018 20:01:33.440944 DEBUG   LWP22584 [net_crypto.c:705 ToxCore] ---- send_packet_to    | PK: f3c923e892339348fbba55d7fde7f368 CRYPTO_DATA
29.04.2018 20:01:33.491148 DEBUG   LWP22584 [net_crypto.c:2525 ToxCore] --- udp_handle_packet | PK: f3c923e892339348fbba55d7fde7f368 CRYPTO_DATA 2
29.04.2018 20:01:33.541524 DEBUG   LWP22584 [net_crypto.c:705 ToxCore] ---- send_packet_to    | PK: f3c923e892339348fbba55d7fde7f368 CRYPTO_DATA
29.04.2018 20:01:33.591672 DEBUG   LWP22584 [net_crypto.c:2525 ToxCore] --- udp_handle_packet | PK: f3c923e892339348fbba55d7fde7f368 CRYPTO_DATA 2
29.04.2018 20:01:33.641851 DEBUG   LWP22584 [net_crypto.c:2525 ToxCore] --- udp_handle_packet | PK: 4b1c717470f18181d00070da1a72f0fe CRYPTO_DATA 2
29.04.2018 20:01:33.641876 DEBUG   LWP22584 [net_crypto.c:2525 ToxCore] --- udp_handle_packet | PK: 43f22f1a9069eb041f5b11b4b6420e64 CRYPTO_DATA 2
29.04.2018 20:01:33.642192 DEBUG   LWP22584 [net_crypto.c:705 ToxCore] ---- send_packet_to    | PK: 4b1c717470f18181d00070da1a72f0fe CRYPTO_DATA
29.04.2018 20:01:34.294424 DEBUG   LWP22584 [net_crypto.c:705 ToxCore] ---- send_packet_to    | PK: 43f22f1a9069eb041f5b11b4b6420e64 CRYPTO_DATA
29.04.2018 20:01:34.394981 DEBUG   LWP22584 [net_crypto.c:705 ToxCore] ---- send_packet_to    | PK: 76518406f6a9f2217e8dc487cc783c25 CRYPTO_DATA
29.04.2018 20:01:34.567944 DEBUG   LWP22584 [net_crypto.c:2525 ToxCore] --- udp_handle_packet | PK: f3c923e892339348fbba55d7fde7f368 CRYPTO_DATA 2
29.04.2018 20:01:34.568022 DEBUG   LWP22584 [net_crypto.c:705 ToxCore] ---- send_packet_to    | PK: f3c923e892339348fbba55d7fde7f368 CRYPTO_DATA
29.04.2018 20:01:34.668285 DEBUG   LWP22584 [net_crypto.c:2525 ToxCore] --- udp_handle_packet | PK: 4b1c717470f18181d00070da1a72f0fe CRYPTO_DATA 2
29.04.2018 20:01:34.668310 DEBUG   LWP22584 [net_crypto.c:2525 ToxCore] --- udp_handle_packet | PK: 43f22f1a9069eb041f5b11b4b6420e64 CRYPTO_DATA 2
29.04.2018 20:01:34.668379 DEBUG   LWP22584 [net_crypto.c:705 ToxCore] ---- send_packet_to    | PK: 4b1c717470f18181d00070da1a72f0fe CRYPTO_DATA
29.04.2018 20:01:35.320351 DEBUG   LWP22584 [net_crypto.c:705 ToxCore] ---- send_packet_to    | PK: 43f22f1a9069eb041f5b11b4b6420e64 CRYPTO_DATA
29.04.2018 20:01:35.572193 DEBUG   LWP22584 [net_crypto.c:705 ToxCore] ---- send_packet_to    | PK: f3c923e892339348fbba55d7fde7f368 CRYPTO_DATA
29.04.2018 20:01:35.622575 DEBUG   LWP22584 [net_crypto.c:2525 ToxCore] --- udp_handle_packet | PK: f3c923e892339348fbba55d7fde7f368 CRYPTO_DATA 2
29.04.2018 20:01:35.672736 DEBUG   LWP22584 [net_crypto.c:2525 ToxCore] --- udp_handle_packet | PK: 4b1c717470f18181d00070da1a72f0fe CRYPTO_DATA 2
29.04.2018 20:01:35.672762 DEBUG   LWP22584 [net_crypto.c:2525 ToxCore] --- udp_handle_packet | PK: 43f22f1a9069eb041f5b11b4b6420e64 CRYPTO_DATA 2
29.04.2018 20:01:35.672831 DEBUG   LWP22584 [net_crypto.c:705 ToxCore] ---- send_packet_to    | PK: 4b1c717470f18181d00070da1a72f0fe CRYPTO_DATA
29.04.2018 20:01:36.275066 DEBUG   LWP22584 [net_crypto.c:2525 ToxCore] --- udp_handle_packet | PK: 76518406f6a9f2217e8dc487cc783c25 CRYPTO_DATA 2
29.04.2018 20:01:36.325285 DEBUG   LWP22584 [net_crypto.c:705 ToxCore] ---- send_packet_to    | PK: 43f22f1a9069eb041f5b11b4b6420e64 CRYPTO_DATA
29.04.2018 20:01:36.325321 DEBUG   LWP22584 [net_crypto.c:705 ToxCore] ---- send_packet_to    | PK: 76518406f6a9f2217e8dc487cc783c25 CRYPTO_DATA
```

**Comment by zoff99** [comment]
@hkarel you can test it yourself. the running echobot should be exactly this code:
https://github.com/alexbakker/EchoBot
ask on the #toktok channel what toxcore version its linked against (i think Impyy is running it)

**Comment by hkarel** [comment]
> use both, while the condition probably fixes the problem most of the time, we should also fix it for very slow connections.

It is made

**Comment by hkarel** [comment]
> @hkarel you can test it yourself. the running echobot should be exactly this code:
https://github.com/alexbakker/EchoBot
ask on the #toktok channel what toxcore version its linked against (i think Impyy is running it)

1. Yes, I can run EchoBot myself, only it will not be high-loaded. All the same, EchoBot 76518406 more than 1000 friends and about 100 simultaneously connected
2. The version of toxcore can be recognized by the command !info
3. I will try to contact Impyy...

**Comment by hkarel** [comment]
> did you try to additionally measure packet loss to echobot?

Do you mean whether all my packages have reached echobot? I do not really know how to do this. 
Yes, and that makes sense? For my client, the indicator of the "established" UDP connection  is determined only coming packets from the EchoBot. Or am I wrong?

**Comment by sudden6** [comment]
no I don't think that's even possible, I meant with a tool like `ping`

**Comment by hkarel** [comment]
In the context of solving this problem, I need to hang up the second handler on  NET_PACKET_GET_NODES command.
I ask the main developers of toxcore to suggest a way how this can be done

**Comment by sudden6** [comment]
@hkarel I don't really understand what you mean, can you maybe reword that sentence?

**Comment by hkarel** [comment]
> @hkarel I don't really understand what you mean, can you maybe reword that sentence?

Now we have the NET_PACKET_GET_NODES command associated with only one handler. 
See DHT.c:2702
`networking_registerhandler(dht->net, NET_PACKET_GET_NODES, &handle_getnodes, dht);`

I would like to have a second handler for the NET_PACKET_GET_NODES command in the net_crypto.c module in order to be able to reset the timestamp (see usage conn->direct_lastrecv_timev4 / conn->direct_lastrecv_timev6 in handler udp_handle_packet()) when a new UDP packet arrives.

This task is interesting in itself, as in "C" you can usually assign only one handler. I would be interested to know the opinion of the Specialists on "C" how to assign two or more handlers.

**Comment by hkarel** [comment]
Over the past week, I continued to investigate this problem.
@irungentoo suggested for the analyze of problem make a dump of network traffic. This gave new food for thought.

Variant 1.
To maintain the status of the UDP connection it is possible to use incoming UDP packets with the commands NET_PACKET_GET_NODES, NET_PACKET_SEND_NODES_IPV6. But for it's will need to add additional handlers for these commands (about this I asked the post above). In addition, these commands do not come often enough. So I decided to abandon this decision.

Variant 2.
I was thinking about the delay in 9 seconds. I noticed that gaps occur when UDP packets of small size (32 bytes) are sent one after another. This is an empty PACKET_ID_REQUEST command.
Then I assumed that some one of the routers holds sequential one-type small packets, and instead of sending one packet per second transmits only one packet in 8 seconds. Perhaps this behavior of the router is due to mechanisms of protection against DDoD attacks.
In order to test this hypothesis, need to send the PACKET_ID_REQUEST command every time with different sizes. To do this within the framework of the current implementation is not possible, in any case I have not found a way.
Perhaps in the future this shortcoming of the tox-network can be used to limit or block it.
As a solution option - we can enter a new command similar to PACKET_ID_REQUEST. But so that this new command performs only one action - the "support" of the UDP connection in the current state, and would not be a dual-use command as now is PACKET_ID_REQUEST.

**Comment by hkarel** [comment]
What in the end?
Now I propose to dwell on the solution that was proposed at the very beginning, only slightly changing it.
Now the parameters MAX_NUM_SENDPACKET_TRIES and UDP_DIRECT_TIMEOUT are interconnected. 
It is how correct to connect parameters of various measurement?  Parameter MAX_NUM_SENDPACKET_ TRIES defines the number of attempts, but the parameter UDP_DIRECT_TIMEUT defines the time interval. I suggest breaking the existing dependency and making the parameter UDP_DIRECT_TIMEUT independent. Like this: 
`#define UDP_DIRECT_TIMEOUT 10` 
This solution is implemented in my last commit. 
What do you think?

**Comment by sudden6** [comment]
> I would like to have a second handler for the NET_PACKET_GET_NODES command in the net_crypto.c

If you need a second handler, you probably need to write your own handler that then calls both sub-handlers. In C these handlers are just a function pointer, which can only accept one function.

About your theory that the packets are buffered, I think that could be possible. Do you know if we maybe need to flush the send buffer after sending such a keep alive packet?

**Comment by hkarel** [comment]
Added mechanism for sending PACKET_ID_KEEPALIVE packets

**Comment by robinlinden** [comment]
Review status: 0 of 3 files reviewed at latest revision, all discussions resolved, some commit checks failed.

---

*[toxcore/net_crypto.c, line 959 at r4](https://reviewable.io/reviews/toktok/c-toxcore/884#-LCFi_cE4E65WK3T5Zlv:-LCFi_cE4E65WK3T5Zlw:b4hnifm) ([raw file](https://github.com/toktok/c-toxcore/blob/727b8e7a5f2f1b5f87a7f533311e078196af47af/toxcore/net_crypto.c#L959)):*
> ```C
> 
>     uint32_t random_len = random_u32() % MAX_SIZE_KEEPALIVE_PACKET;
>     randombytes(data + 1, random_len);
> ```

Use toxcore/crypto_core's `random_bytes` instead of directly using libsodium.

---


*Comments from [Reviewable](https://reviewable.io/reviews/toktok/c-toxcore/884)*

**Comment by hkarel** [comment]
> Use toxcore/crypto_core's random_bytes instead of directly using libsodium. 

Ok, I'll take this into consideration

**Comment by hkarel** [comment]
The last three days I have been testing the last solution. During this time, there was no "ping-pong" incidents of TCP/UDP switching. 
I ask to accept these changes.

**Comment by sudden6** [comment]
Can you provide a reason for the change in L2541 (return -> continue)?

Else the change lgtm

**Comment by hkarel** [comment]
> Can you provide a reason for the change in L2541 (return -> continue)?

@sudden6 , you probably meant L2451.
Now how did I come to this: After I added the command PACKET_ID_KEEPALIVE, the amount of ping-pong TCP/UDP incidents dropped dramatically, but it did not disappear at all (as I now understand, it was just a coincidence). Next, I figured out why every 9th second the connection again switches to the UDP.  It turned that at module friend_connection.c has a function send_pings(), which is called on the 9th second. Further, the incident with the "loss" of UDP packets happened on my local network. I thought it was strange: from one point (func. send_pings()) packets always come, from another point (func. send_crypto_packets()) - not always, and this happens on the local network. Next, I assumed that the send_request_packet() function is not always called, and started looking for a reason. The only thing I found was the condition:
```
        if (conn == nullptr) {
            return;
        }
```
Then I wrote a letter to @irungentoo  and asked him why the condition is exactly this. He replied that it is most likely a mistake. I made corrections and started testing. 
By the way, in the code I found at least two more places, where there is the same error.

**Comment by hkarel** [comment]
P.S. 
I want to tell thanks of Impyy for the help in testing (he had to repeatedly rebuild the EchoBot), without him I couldn't eliminate an error.
And also to thank @irungentoo for the fact that responded to my letters

**Comment by sudden6** [comment]
:lgtm_strong: thanks for this detailed explaination

---

Review status: 0 of 2 files reviewed at latest revision, 1 unresolved discussion.

---



*Comments from [Reviewable](https://reviewable.io/reviews/toktok/c-toxcore/884#-:-LCnSObjC84MYfwYKhql:beimjl3)*

**Comment by sudden6** [comment]
Reviewed 2 of 3 files at r5.
Review status: all files reviewed at latest revision, 1 unresolved discussion, all commit checks successful.

---



*Comments from [Reviewable](https://reviewable.io/reviews/toktok/c-toxcore/884)*

**Comment by hkarel** [comment]
*[toxcore/net_crypto.c, line 959 at r4](https://reviewable.io/reviews/toktok/c-toxcore/884#-LCFi_cE4E65WK3T5Zlv:-LD18CD47VQImaa_5iJu:b-6mgeuk) ([raw file](https://github.com/toktok/c-toxcore/blob/727b8e7a5f2f1b5f87a7f533311e078196af47af/toxcore/net_crypto.c#L959)):*
<details><summary><i>Previously, robinlinden (Robin Lindén) wrote…</i></summary><blockquote>

Use toxcore/crypto_core's `random_bytes` instead of directly using libsodium.
</blockquote></details>

Ok, I'll take this into account in the future. For the latest implementation, this observation has become irrelevant.

---


*Comments from [Reviewable](https://reviewable.io/reviews/toktok/c-toxcore/884)*

**Comment by iphydf** [comment]
This is a 4 line PR made of 3 commits. Can you turn it into a single commit with an appropriate description?

**Comment by hkarel** [comment]
I have tried to make as you asked. Look please.

---







*Comments from [Reviewable](https://reviewable.io/reviews/toktok/c-toxcore/884#-:-LDCQaIq7oyknSSoPErt:b-cr709v)*

**Comment by hkarel** [comment]
Next, I plan to do another PR. In it there will be two small changes, I found them in result of work on this problem, but did not merge into one PR, because will probably need to testing.

---







*Comments from [Reviewable](https://reviewable.io/reviews/toktok/c-toxcore/884#-:-LDCS1858aExsA51jJUq:b-ftzxns)*
